### ISSUE TITLE
Misleading error : Neither type `Function` nor type `SomeClass` is assignable to the other

### ISSUE THREAD

**Issue by basarat** [Microsoft/TypeScript#5445]
Code:

```ts
class SomeClass {
    method() {
		(this.constructor as SomeClass); // Error here
    }
}

var foo:Function;
foo = SomeClass; // No Error: They are assignable

```
# Screenshot

![image](https://cloud.githubusercontent.com/assets/874898/10807860/ce2015f4-7e38-11e5-8390-438ab4cb9159.png)

# PS 
The following compiles fine (i.e. if you move that to the constructor body)

```ts
class SomeClass {
    constructor(){
		(this.constructor as SomeClass); // No Error
	}
}

var foo:Function;
foo = SomeClass; // No Error: They are assignable
```

:rose:

**Comment by basarat** [comment]
Sorry. Instances are not same as functions (doh). So the place where it doesn't error is the actual error. I'll create a separate issue. Ignore this one :rose:

**Comment by basarat** [comment]
> So the place where it doesn't error is the actual error

And even that's not a bug ... Its just that an empty object (SomeClass instances have no properties in that case) is assignable. 

So move along. Nothing to see here :rose:

**Comment by bobbyinsane** [comment]
@basarat 

```typescript
class Shape<T extends Shape<any>>{
    constructor(){
        console.info("Shape..");
    }
    public draw():T{
        return <T>this;
    }

    public addStroke():T{
        return <T>this;
    }
}
```
 this code works perfectly in v1.6.2, and throwing 
<pre>
Error:(10, 16) TS2352: Neither type 'this' nor type 'T' is assignable to the other.
 Type 'Shape<T>' is not assignable to type 'T'. </pre> with v1.7.3, Any idea how to resolve this?

**Comment by basarat** [comment]
> Neither type 'this' nor type 'T' is assignable to the other.

That error can be silenced with a double assertion: 

```ts
class Shape<T extends Shape<any>>{
    constructor(){
        console.info("Shape..");
    }
    public draw():T{
        return <T><any>this;
    }

    public addStroke():T{
        return <T><any>this;
    }
} 
```

https://basarat.gitbooks.io/typescript/content/docs/types/type-assertion.html

**Comment by bobbyinsane** [comment]
it worked. Now the problem is, the IDE that I'm using stopped showing the suggestions(IntelliSense). Basically, I wanted to have method chaining and with earlier code I was able to get IntelliSense, but after double assertion, it stopped showing the parent class methods in the child classes. Is it the compiler problem? or side effects of 'double assertion'.?

**Comment by salihkesepara** [comment]
test
